apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl

```diff
@@ -1,277 +1,253 @@
--export_type([db_options/0, db_write_options/0, db_read_options/0]).
-
--compile({inline, [meta_lookup/2]}).
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
 
 %%================================================================================
 %% Type declarations
 %%================================================================================
 
--type options() :: #{
-    dir => file:filename()
-}.
+-type prototype() ::
+    {emqx_ds_storage_reference, emqx_ds_storage_reference:options()}
+    | {emqx_ds_storage_bitfield_lts, emqx_ds_storage_bitfield_lts:options()}.
 
-%% see rocksdb:db_options()
--type db_options() :: proplists:proplist().
-%% see rocksdb:write_options()
--type db_write_options() :: proplists:proplist().
-%% see rocksdb:read_options()
--type db_read_options() :: proplists:proplist().
+-type shard_id() :: emqx_ds_replication_layer:shard_id().
 
 -type cf_refs() :: [{string(), rocksdb:cf_handle()}].
 
-%% Message storage generation
-%% Keep in mind that instances of this type are persisted in long-term storage.
--type generation() :: #{
-    %% Module that handles data for the generation
-    module := module(),
-    %% Module-specific data defined at generation creation time
-    data := term(),
-    %% When should this generation become active?
-    %% This generation should only contain messages timestamped no earlier than that.
-    %% The very first generation will have `since` equal 0.
-    since := emqx_ds:time()
-}.
+-type gen_id() :: 0..16#ffff.
 
--record(s, {
-    shard :: emqx_ds:shard(),
-    db :: rocksdb:db_handle(),
-    cf_iterator :: rocksdb:cf_handle(),
-    cf_generations :: cf_refs()
+%% Note: this record might be stored permanently on a remote node.
+-record(stream, {
+    generation :: gen_id(),
+    enc :: _EncapsultatedData,
+    misc = #{} :: map()
 }).
 
+-opaque stream() :: #stream{}.
+
+%% Note: this record might be stored permanently on a remote node.
 -record(it, {
-    shard :: emqx_ds:shard(),
-    gen :: gen_id(),
-    replay :: emqx_ds:replay(),
-    module :: module(),
-    data :: term()
+    generation :: gen_id(),
+    enc :: _EncapsultatedData,
+    misc = #{} :: map()
 }).
 
--type gen_id() :: 0..16#ffff.
-
--opaque state() :: #s{}.
 -opaque iterator() :: #it{}.
 
-%% Contents of the default column family:
-%%
-%% [{<<"genNN">>, #generation{}}, ...,
-%%  {<<"current">>, GenID}]
+%%%% Generation:
 
--define(DEFAULT_CF, "default").
--define(DEFAULT_CF_OPTS, []).
+-type generation(Data) :: #{
+    %% Module that handles data for the generation:
+    module := module(),
+    %% Module-specific data defined at generation creation time:
+    data := Data,
+    %% When should this generation become active?
+    %% This generation should only contain messages timestamped no earlier than that.
+    %% The very first generation will have `since` equal 0.
+    since := emqx_ds:time(),
+    until := emqx_ds:time() | undefined
```

Comment:
Nit. This is a redundant piece of information, shouldn't `since` be enough?

Reply:
I think this piece of information will be useful in the future.